packages/modules/kostal_plenticore/device.py

```diff
@@ -0,0 +1,165 @@
+        components: Iterable[Union[KostalPlenticoreBat, KostalPlenticoreCounter, KostalPlenticoreInverter]]
+    ):
+        with tcp_client:
+            update(components, reader)
+
+    tcp_client = modbus.ModbusTcpClient_(device_config.configuration.ip_address, 1502)
+    reader = _create_reader(tcp_client)
+    return ConfigurableDevice(
+        device_config=device_config,
+        component_factory=ComponentFactoryByType(
+            bat=create_bat_component, counter=create_counter_component, inverter=create_inverter_component),
+        component_updater=MultiComponentUpdater(update_components),
+    )
+
+
+def _create_reader(tcp_client: modbus.ModbusTcpClient_) -> Callable[[int, modbus.ModbusDataType], Any]:
+    return functools.partial(tcp_client.read_holding_registers, unit=71, wordorder=Endian.Little)
+
+
+def read_legacy_inverter(ip1: str, ip2: str, battery: int, ip3: str) -> InverterState:
+    # in IP3 kann ein aufeinanderfolgende Liste enthalten sein "192.168.0.1-3"
+    log.debug("Kostal Plenticore: WR1: %s, WR2: %s, WR3: %s, Battery: %s", ip1, ip2, ip3, battery)
+    inverter_component = KostalPlenticoreInverter(KostalPlenticoreInverterSetup(id=1))
+
+    def get_hybrid_inverter_state(ip: str) -> InverterState:
+        battery_component = KostalPlenticoreBat(1, KostalPlenticoreBatSetup())
+        with modbus.ModbusTcpClient_(ip) as client:
+            return update(
+                [inverter_component, battery_component], _create_reader(client), set_inverter_state=False
+            )
+
+    def get_standard_inverter_state(ip: str) -> InverterState:
+        with modbus.ModbusTcpClient_(ip) as client:
+            return inverter_component.update(_create_reader(client))
+
+    def inverter_state_sum(a: InverterState, b: InverterState) -> InverterState:
+        return InverterState(exported=a.exported + b.exported, power=a.power + b.power)
+
+    inverter_state = functools.reduce(
+        inverter_state_sum,
+        map(get_standard_inverter_state, filter("none".__ne__, chain([ip2], ipparser(ip3)))),
+        get_hybrid_inverter_state(ip1) if battery else get_standard_inverter_state(ip1)
+    )
+    inverter_component.set(inverter_state)
+    return inverter_state
+
+
+def read_legacy_counter(ip1: str, ip2: str, battery: int, ip3: str, position: int) -> None:
+    log.debug("Kostal Plenticore: WR1: %s, Position: %s", ip1, position)
+    client = modbus.ModbusTcpClient_(ip1, 1502)
+    reader = _create_reader(client)
+    counter_component = KostalPlenticoreCounter(None, KostalPlenticoreCounterSetup(id=None))
+    if LegacyCounterPosition(position) == LegacyCounterPosition.GRID:
+        with client:
+            counter_component.update(reader)
+    else:
+        with client:
+            counter_state = counter_component.get_values(reader)
+            bat_power = KostalPlenticoreBat(None, KostalPlenticoreBatSetup()).update(reader).power
+        inverter_power = read_legacy_inverter(ip1, ip2, battery, ip3).power
```

Comment:
In dem Fall würde das Update zwei mal durchgeführt werden. Schlimm ist das vermutlich nicht. Wahrscheinlich haben das Setup so sowieso nur 3 Leute ;-). Es könnte aber ein Gedanke wert sein, den Teil auch an die "große" Abfrage anzuhängen.